File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/photoInfo/PhotoInfoFragment.java
package com.mudrichenko.evgeniy.flickrtestproject.ui.photoInfo;

import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.mudrichenko.evgeniy.flickrtestproject.App;
import com.mudrichenko.evgeniy.flickrtestproject.R;
import com.mudrichenko.evgeniy.flickrtestproject.database.FlickrPhoto;
import com.mudrichenko.evgeniy.flickrtestproject.ui.BaseFragment;
import com.mudrichenko.evgeniy.flickrtestproject.utils.PrefUtils;
import com.mudrichenko.evgeniy.flickrtestproject.utils.StringUtils;
import com.todddavies.components.progressbar.ProgressWheel;

import javax.inject.Inject;

public class PhotoInfoFragment extends BaseFragment implements PhotoInfoView, View.OnClickListener {

    @InjectPresenter
    PhotoInfoPresenter mPhotoInfoPresenter;

    @Inject
    StringUtils mStringUtils;

    @Inject
    PrefUtils mPrefUtils;

    PublicPhotoPresenterListener mPublicPhotoPresenterListener;

    private long mPhotoId;

    private RelativeLayout mInfoLayout;
    private ProgressWheel mProgressWheel;

    private ImageView mButtonViewClose;
    private ImageView mButtonViewDelete;

    TextView mTextViewInfo;

    private TextView mTextViewPhotoName;
    private TextView mTextViewDescription;
    private TextView mTextViewLocation;
    private TextView mTextViewTakenBy;
    private TextView mTextViewDateTaken;
    private TextView mTextViewTags;
    private TextView mTextViewLicense;
    private TextView mTextViewPrivacy;
    private TextView mTextViewViewsCount;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_photo_info, container, false);
        App.Companion.getAppComponent().inject(this);
        view.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                return true;
            }
        });
        mInfoLayout = view.findViewById(R.id.infoLayout);
        mProgressWheel = view.findViewById(R.id.progress_wheel);
        mTextViewInfo = view.findViewById(R.id.text_view_info);
        mButtonViewClose = view.findViewById(R.id.btnViewClose);
        mButtonViewClose.setOnClickListener(this);
        mButtonViewDelete = view.findViewById(R.id.btnViewDelete);
        mButtonViewDelete.setOnClickListener(this);
        mButtonViewDelete.setVisibility(View.INVISIBLE);
        mTextViewPhotoName = view.findViewById(R.id.textViewPhotoName);
        mTextViewDescription = view.findViewById(R.id.textViewDescription);
        mTextViewLocation = view.findViewById(R.id.textViewLocation);
        mTextViewTakenBy = view.findViewById(R.id.textViewTakenBy);
        mTextViewDateTaken = view.findViewById(R.id.textViewDateTaken);
        mTextViewTags = view.findViewById(R.id.textViewTags);
        mTextViewLicense = view.findViewById(R.id.textViewLicense);
        mTextViewPrivacy = view.findViewById(R.id.textViewPrivacy);
        mTextViewViewsCount = view.findViewById(R.id.textViewViewsCount);
        if (getArguments() != null) {
            mPhotoId = getArguments().getLong("mPhotoId");
        }
        mPhotoInfoPresenter.onPhotoInfoLoad(mPhotoId);
        return view;
    }

    public static PhotoInfoFragment newInstance(Long mPhotoId) {
        PhotoInfoFragment fragment = new PhotoInfoFragment();
        Bundle bundle = new Bundle();
        bundle.putLong("mPhotoId", mPhotoId);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void showPhotoInfo(FlickrPhoto flickrPhoto) {
        String mPhotoOwnerNSID = flickrPhoto.getOwnerId();
        if (mPhotoOwnerNSID != null) {
            if (mPhotoOwnerNSID.equals(mPrefUtils.getUserId())) {
                mButtonViewDelete.setVisibility(View.VISIBLE);
            }
        }
        mTextViewPhotoName.setText(flickrPhoto.getName());
        mTextViewDescription.setText(flickrPhoto.getDescription());
        //mTextViewLocation.setText(mStringUtils.getLocationName(flickrPhoto.locgetLocation()));
        mTextViewTakenBy.setText(mStringUtils.getOwnerName(flickrPhoto.getOwnerName(), flickrPhoto.getOwnerRealName()));
        mTextViewDateTaken.setText(flickrPhoto.getDateTaken());
        //mTextViewTags.setText(mStringUtils.getTagsString(flickrPhoto.getTags().getTag()));
        //mTextViewLicense.setText(mStringUtils.getLicenseStringById(flickrPhoto.getLicense()));
        mTextViewPrivacy.setText(mStringUtils.getPrivacyStringById(flickrPhoto.getSafetyLevel()));
        mTextViewViewsCount.setText(String.valueOf(flickrPhoto.getNumOfViews()));
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPhotoInfoPresenter.unsubscribe();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPhotoInfoPresenter.unsubscribe();
    }

    @Override
    public void showProgressWheel() {
        mInfoLayout.setVisibility(View.VISIBLE);
        mTextViewInfo.setVisibility(View.INVISIBLE);
        mProgressWheel.setVisibility(View.VISIBLE);
        mProgressWheel.startSpinning();
    }

    @Override
    public void hideProgressWheel() {
        mInfoLayout.setVisibility(View.INVISIBLE);
        mProgressWheel.setVisibility(View.INVISIBLE);
        mProgressWheel.stopSpinning();
    }

    @Override
    public void showInfoMessage(String message) {
        mInfoLayout.setVisibility(View.VISIBLE);
        mProgressWheel.setVisibility(View.INVISIBLE);
        mTextViewInfo.setVisibility(View.VISIBLE);
        mTextViewInfo.setText(message);
    }

    @Override
    public void hideInfoMessage() {
        mInfoLayout.setVisibility(View.INVISIBLE);
        mTextViewInfo.setVisibility(View.INVISIBLE);
    }

    @Override
    public void onPhotoDeleted() {
        if (mPublicPhotoPresenterListener != null) {
            mPublicPhotoPresenterListener.onPhotoDeleted();
        }
        removeFragment();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btnViewClose:
                clickOnBtnClose();
                break;
            case R.id.btnViewDelete:
                clickOnBtnDelete();
                break;
        }
    }

    private void clickOnBtnClose() {
        removeFragment();
    }

    private void removeFragment() {
        FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
        fragmentManager.beginTransaction().remove(this).commit();
        fragmentManager.popBackStack();
    }

    private void clickOnBtnDelete() {
        mPhotoInfoPresenter.onButtonDeleteClick(mPhotoId);
    }

    public interface PublicPhotoPresenterListener {

        void onPhotoDeleted();

    }

    public void setPublicPhotoPresenterListener(PublicPhotoPresenterListener publicPhotoPresenterListener) {
        mPublicPhotoPresenterListener = publicPhotoPresenterListener;
    }


}

File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/photoInfo/PhotoInfoFragment.kt
package com.mudrichenko.evgeniy.flickrtestproject.ui.photoInfo

import android.os.Bundle
import androidx.fragment.app.FragmentManager
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView

import com.arellomobile.mvp.presenter.InjectPresenter
import com.mudrichenko.evgeniy.flickrtestproject.App
import com.mudrichenko.evgeniy.flickrtestproject.R
import com.mudrichenko.evgeniy.flickrtestproject.database.FlickrPhoto
import com.mudrichenko.evgeniy.flickrtestproject.ui.BaseFragment
import com.mudrichenko.evgeniy.flickrtestproject.utils.PrefUtils
import com.mudrichenko.evgeniy.flickrtestproject.utils.StringUtils
import com.todddavies.components.progressbar.ProgressWheel

import javax.inject.Inject

class PhotoInfoFragment : BaseFragment(), PhotoInfoView, View.OnClickListener {

    @InjectPresenter
    internal lateinit var mPhotoInfoPresenter: PhotoInfoPresenter

    @Inject
    internal lateinit var mStringUtils: StringUtils

    @Inject
    internal lateinit var mPrefUtils: PrefUtils

    internal lateinit var mPublicPhotoPresenterListener: PublicPhotoPresenterListener

    private var mPhotoId: Long = 0

    private lateinit var mInfoLayout: RelativeLayout
    private lateinit var mProgressWheel: ProgressWheel

    private lateinit var mButtonViewClose: ImageView
    private lateinit var mButtonViewDelete: ImageView

    internal lateinit var mTextViewInfo: TextView

    private var mTextViewPhotoName: TextView? = null
    private var mTextViewDescription: TextView? = null
    private var mTextViewLocation: TextView? = null
    private var mTextViewTakenBy: TextView? = null
    private var mTextViewDateTaken: TextView? = null
    private var mTextViewTags: TextView? = null
    private var mTextViewLicense: TextView? = null
    private var mTextViewPrivacy: TextView? = null
    private var mTextViewViewsCount: TextView? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_photo_info, container, false)
        App.appComponent!!.inject(this)
        view.setOnTouchListener { view, motionEvent -> true }
        mInfoLayout = view.findViewById(R.id.infoLayout)
        mProgressWheel = view.findViewById(R.id.progress_wheel)
        mTextViewInfo = view.findViewById(R.id.text_view_info)
        mButtonViewClose = view.findViewById(R.id.btnViewClose)
        mButtonViewClose.setOnClickListener(this)
        mButtonViewDelete = view.findViewById(R.id.btnViewDelete)
        mButtonViewDelete.setOnClickListener(this)
        mButtonViewDelete.visibility = View.INVISIBLE
        mTextViewPhotoName = view.findViewById(R.id.textViewPhotoName)
        mTextViewDescription = view.findViewById(R.id.textViewDescription)
        mTextViewLocation = view.findViewById(R.id.textViewLocation)
        mTextViewTakenBy = view.findViewById(R.id.textViewTakenBy)
        mTextViewDateTaken = view.findViewById(R.id.textViewDateTaken)
        mTextViewTags = view.findViewById(R.id.textViewTags)
        mTextViewLicense = view.findViewById(R.id.textViewLicense)
        mTextViewPrivacy = view.findViewById(R.id.textViewPrivacy)
        mTextViewViewsCount = view.findViewById(R.id.textViewViewsCount)
        if (arguments != null) {
            mPhotoId = arguments!!.getLong("mPhotoId")
        }
        mPhotoInfoPresenter.onPhotoInfoLoad(mPhotoId)
        return view
    }

    override fun showPhotoInfo(flickrPhoto: FlickrPhoto) {
        val mPhotoOwnerNSID = flickrPhoto.ownerId
        if (mPhotoOwnerNSID != null) {
            if (mPhotoOwnerNSID == mPrefUtils.getUserId()) {
                mButtonViewDelete.visibility = View.VISIBLE
            }
        }
        mTextViewPhotoName!!.text = flickrPhoto.name
        mTextViewDescription!!.text = flickrPhoto.description
        //mTextViewLocation.setText(mStringUtils.getLocationName(flickrPhoto.locgetLocation()));
        mTextViewTakenBy!!.text = mStringUtils.getOwnerName(flickrPhoto.ownerName, flickrPhoto.ownerRealName)
        mTextViewDateTaken!!.text = flickrPhoto.dateTaken
        //mTextViewTags.setText(mStringUtils.getTagsString(flickrPhoto.getTags().getTag()));
        //mTextViewLicense.setText(mStringUtils.getLicenseStringById(flickrPhoto.getLicense()));
        mTextViewPrivacy!!.text = mStringUtils.getPrivacyStringById(flickrPhoto.safetyLevel)
        mTextViewViewsCount!!.text = flickrPhoto.numOfViews.toString()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPhotoInfoPresenter.unsubscribe()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mPhotoInfoPresenter.unsubscribe()
    }

    override fun showProgressWheel() {
        mInfoLayout.visibility = View.VISIBLE
        mTextViewInfo.visibility = View.INVISIBLE
        mProgressWheel.visibility = View.VISIBLE
        mProgressWheel.startSpinning()
    }

    override fun hideProgressWheel() {
        mInfoLayout.visibility = View.INVISIBLE
        mProgressWheel.visibility = View.INVISIBLE
        mProgressWheel.stopSpinning()
    }

    override fun showInfoMessage(message: String) {
        mInfoLayout.visibility = View.VISIBLE
        mProgressWheel.visibility = View.INVISIBLE
        mTextViewInfo.visibility = View.VISIBLE
        mTextViewInfo.text = message
    }

    override fun hideInfoMessage() {
        mInfoLayout.visibility = View.INVISIBLE
        mTextViewInfo.visibility = View.INVISIBLE
    }

    override fun onPhotoDeleted() {
        mPublicPhotoPresenterListener.onPhotoDeleted()
        removeFragment()
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btnViewClose -> clickOnBtnClose()
            R.id.btnViewDelete -> clickOnBtnDelete()
        }
    }

    private fun clickOnBtnClose() {
        removeFragment()
    }

    private fun removeFragment() {
        val fragmentManager = activity!!.supportFragmentManager
        fragmentManager.beginTransaction().remove(this).commit()
        fragmentManager.popBackStack()
    }

    private fun clickOnBtnDelete() {
        mPhotoInfoPresenter.onButtonDeleteClick(mPhotoId)
    }

    interface PublicPhotoPresenterListener {

        fun onPhotoDeleted()

    }

    fun setPublicPhotoPresenterListener(publicPhotoPresenterListener: PublicPhotoPresenterListener) {
        mPublicPhotoPresenterListener = publicPhotoPresenterListener
    }

    companion object {

        fun newInstance(mPhotoId: Long?): PhotoInfoFragment {
            val fragment = PhotoInfoFragment()
            val bundle = Bundle()
            bundle.putLong("mPhotoId", mPhotoId!!)
            fragment.arguments = bundle
            return fragment
        }
    }


}
